sw/host/opentitanlib/src/util/openocd.rs

```diff
@@ -0,0 +1,256 @@
+            .context("failed to send a command to OpenOCD server")?;
+        stream
+            .write_all(&[0x1a])
+            .context("failed to send a command to OpenOCD server")?;
+
+        stream.flush().context("failed to flush stream")?;
+
+        let answer = stream
+            .borrow_mut()
+            .bytes()
+            .take_while(|x| match x {
+                Ok(x) => *x != 0x1a,
+                _ => false,
+            })
+            .collect::<Result<_, _>>()
+            .context("failed to read the result of a command from OpenOCD server")?;
+
+        self.openocd_socket_stream.replace(Some(stream));
+        String::from_utf8(answer).context("failed to parse OpenOCD response as UTF-8")
+    }
+}
+
+impl Drop for OpenOcdServer {
+    fn drop(&mut self) {
+        self.stop().ok();
+    }
+}
+
+impl Jtag for OpenOcdServer {
+    fn connect(&self, tap: JtagTap) -> Result<()> {
+        self.start(tap)
+    }
+
+    fn disconnect(&self) -> Result<()> {
+        self.stop()
+    }
+
+    fn read_lc_ctrl_reg(&self, reg: &LcCtrlReg) -> Result<u32> {
+        let reg_offset = reg.word_offset();
+        let cmd = format!("riscv dmi_read 0x{reg_offset:x}");
+        let response = self.send_tcl_cmd(cmd.as_str())?;
+
+        let value = parse_hex_word(response.trim()).context(format!(
+            "expected response to be hexadecimal word, got '{response}'"
+        ))?;
+
+        Ok(value)
+    }
+
+    fn write_lc_ctrl_reg(&self, reg: &LcCtrlReg, value: u32) -> Result<()> {
+        let reg_offset = reg.word_offset();
+        let cmd = format!("riscv dmi_write 0x{reg_offset:x} 0x{value:x}");
+        let response = self.send_tcl_cmd(cmd.as_str())?;
+
+        if !response.is_empty() {
+            bail!("unexpected response: '{response}'");
+        }
+
+        Ok(())
+    }
+
+    fn read_memory(&self, addr: u32, buf: &mut [u8]) -> Result<u32> {
+        let cmd = format!("read_memory 0x{addr:x} 8 {count}", count = buf.len());
+        let _response = self.send_tcl_cmd(cmd.as_str())?;
+        todo!("handle response from OpenOCD");
+    }
+
+    fn write_memory(&self, addr: u32, buf: &[u8]) -> Result<u32> {
+        // Convert data to space-separated strings
+        let data: Vec<_> = buf.iter().map(ToString::to_string).collect();
+        let data_str = &data[..].join(" ");
+
+        let cmd = format!("write_memory 0x{addr:x} 8 {{ {data_str} }}");
+        let _response = self.send_tcl_cmd(cmd.as_str())?;
+
+        todo!("handle response from OpenOCD");
+    }
+
+    fn halt(&self) -> Result<()> {
+        let _response = self.send_tcl_cmd("halt")?;
+        todo!("handle response from OpenOCD");
+    }
+
+    fn resume(&self) -> Result<()> {
+        let _response = self.send_tcl_cmd("resume")?;
+        todo!("handle response from OpenOCD");
+    }
+}
+
+/// Parse a string as a hexadecimal word.
+///
+/// Expects an input string of the form `0x[0-9a-fA-F]{1,8}`.
+fn parse_hex_word(s: &str) -> Result<u32> {
```

Comment:
Could you not use what is already here: https://cs.opensource.google/opentitan/opentitan/+/master:sw/host/opentitanlib/src/util/parse_int.rs;drc=48f64de05bb50cb37bf2f69a2d33d86fa3de0ddb;l=24

?